File: backend/app/api/routes/debt.py
import calendar
import uuid
from datetime import date
from decimal import Decimal
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Response, status
from sqlalchemy.orm import Session

from app.db.session import get_db_session
from app.schemas.debt import (
    AmortizationPaymentResponse,
    AmortizationProjectionRequest,
    AmortizationProjectionResponse,
    LoanBalanceCreate,
    LoanBalanceResponse,
    LoanBalanceUpdate,
    LoanTermsCreate,
    LoanTermsResponse,
    LoanTermsUpdate,
    ProjectionAssumptions,
    ScenarioComparisonResponse,
    ScenarioDebtPaymentResponse,
    ScenarioMonthResponse,
    ScenarioProjectionRequest,
    ScenarioResultResponse,
)
from app.services import debt, loans

router = APIRouter()
projection_router = APIRouter()
SessionDependency = Annotated[Session, Depends(get_db_session)]


def _loan_not_found() -> HTTPException:
    return HTTPException(status_code=404, detail="Loan not found")


def _balance_not_found() -> HTTPException:
    return HTTPException(status_code=404, detail="Balance snapshot not found")


def _conflict(detail: str) -> HTTPException:
    return HTTPException(status_code=409, detail=detail)


def _projection_error(error: debt.AmortizationError) -> HTTPException:
    return HTTPException(status_code=422, detail=str(error))


def _payment_date(first_payment_date: date, offset: int) -> date:
    month_index = first_payment_date.month - 1 + offset
    year = first_payment_date.year + month_index // 12
    month = month_index % 12 + 1
    day = min(first_payment_date.day, calendar.monthrange(year, month)[1])
    return date(year, month, day)


def _assumptions(max_months: int) -> ProjectionAssumptions:
    return ProjectionAssumptions(
        apr_treatment="Fixed nominal APR where 100 basis points equals 1% APR.",
        periodic_rate="APR divided by 12; no daily-interest calculation.",
        compounding="Interest compounds monthly.",
        payment_timing="Monthly interest accrues before that month's payment.",
        currency_rounding="Interest and balances round to cents using ROUND_HALF_UP.",
        final_payment="The final payment is clamped to principal plus accrued interest.",
        maximum_months=max_months,
        disclaimer="Projections are estimates, not financial advice or payoff guarantees.",
    )


@router.get("", response_model=list[LoanTermsResponse])
def list_all(session: SessionDependency) -> list[LoanTermsResponse]:
    return [LoanTermsResponse.model_validate(item) for item in loans.list_loans(session)]


@router.post("", response_model=LoanTermsResponse, status_code=status.HTTP_201_CREATED)
def create(data: LoanTermsCreate, session: SessionDependency) -> LoanTermsResponse:
    try:
        loan = loans.create_loan(session, data)
    except loans.LoanAccountError as error:
        raise _conflict("Loan terms require an existing loan account") from error
    except loans.LoanConflictError as error:
        raise _conflict("The account already has loan terms") from error
    return LoanTermsResponse.model_validate(loan)


@router.get("/{loan_id}", response_model=LoanTermsResponse)
def get(loan_id: uuid.UUID, session: SessionDependency) -> LoanTermsResponse:
    try:
        loan = loans.get_loan(session, loan_id)
    except loans.LoanNotFoundError as error:
        raise _loan_not_found() from error
    return LoanTermsResponse.model_validate(loan)


@router.patch("/{loan_id}", response_model=LoanTermsResponse)
def update(
    loan_id: uuid.UUID, data: LoanTermsUpdate, session: SessionDependency
) -> LoanTermsResponse:
    try:
        loan = loans.update_loan(session, loan_id, data)
    except loans.LoanNotFoundError as error:
        raise _loan_not_found() from error
    return LoanTermsResponse.model_validate(loan)


@router.delete("/{loan_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete(loan_id: uuid.UUID, session: SessionDependency) -> Response:
    try:
        loans.delete_loan(session, loan_id)
    except loans.LoanNotFoundError as error:
        raise _loan_not_found() from error
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/{loan_id}/balances", response_model=list[LoanBalanceResponse])
def list_balance_history(
    loan_id: uuid.UUID, session: SessionDependency
) -> list[LoanBalanceResponse]:
    try:
        snapshots = loans.list_balances(session, loan_id)
    except loans.LoanNotFoundError as error:
        raise _loan_not_found() from error
    return [LoanBalanceResponse.model_validate(item) for item in snapshots]


@router.post(
    "/{loan_id}/balances",
    response_model=LoanBalanceResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_balance(
    loan_id: uuid.UUID, data: LoanBalanceCreate, session: SessionDependency
) -> LoanBalanceResponse:
    try:
        snapshot = loans.create_balance(session, loan_id, data)
    except loans.LoanNotFoundError as error:
        raise _loan_not_found() from error
    except loans.BalanceConflictError as error:
        raise _conflict("A balance already exists for this loan and date") from error
    return LoanBalanceResponse.model_validate(snapshot)


@router.patch(
    "/{loan_id}/balances/{balance_id}", response_model=LoanBalanceResponse
)
def update_balance(
    loan_id: uuid.UUID,
    balance_id: uuid.UUID,
    data: LoanBalanceUpdate,
    session: SessionDependency,
) -> LoanBalanceResponse:
    try:
        snapshot = loans.update_balance(session, loan_id, balance_id, data)
    except loans.BalanceNotFoundError as error:
        raise _balance_not_found() from error
    except loans.BalanceConflictError as error:
        raise _conflict("A balance already exists for this loan and date") from error
    return LoanBalanceResponse.model_validate(snapshot)


@router.delete(
    "/{loan_id}/balances/{balance_id}", status_code=status.HTTP_204_NO_CONTENT
)
def delete_balance(
    loan_id: uuid.UUID, balance_id: uuid.UUID, session: SessionDependency
) -> Response:
    try:
        loans.delete_balance(session, loan_id, balance_id)
    except loans.BalanceNotFoundError as error:
        raise _balance_not_found() from error
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@projection_router.post(
    "/amortization", response_model=AmortizationProjectionResponse
)
def project_amortization(
    data: AmortizationProjectionRequest,
) -> AmortizationProjectionResponse:
    try:
        schedule = debt.calculate_amortization(
            principal=data.principal,
            annual_rate_basis_points=data.annual_rate_basis_points,
            monthly_payment=data.monthly_payment,
            max_months=data.max_months,
        )
    except debt.AmortizationError as error:
        raise _projection_error(error) from error

    payments = [
        AmortizationPaymentResponse(
            month=row.month,
            payment_date=_payment_date(data.first_payment_date, row.month - 1),
            starting_balance=row.starting_balance,
            interest=row.interest,
            payment=row.payment,
            principal=row.principal,
            ending_balance=row.ending_balance,
        )
        for row in schedule.payments
    ]
    return AmortizationProjectionResponse(
        assumptions=_assumptions(data.max_months),
        payments=payments,
        total_interest=schedule.total_interest,
        total_paid=schedule.total_paid,
        months=schedule.months,
        payoff_date=payments[-1].payment_date if payments else None,
        annual_rate_basis_points=schedule.annual_rate_basis_points,
        monthly_rate=schedule.monthly_rate,
    )


def _scenario_response(
    scenario: debt.PayoffScenario,
    first_payment_date: date,
    *,
    baseline_months: int,
    baseline_interest: Decimal,
) -> ScenarioResultResponse:
    schedule = [
        ScenarioMonthResponse(
            month=row.month,
            payment_date=_payment_date(first_payment_date, row.month - 1),
            extra_payment_targets=list(row.extra_payment_targets),
            payments=[
                ScenarioDebtPaymentResponse(
                    debt_id=payment.debt_id,
                    starting_balance=payment.starting_balance,
                    interest=payment.interest,
                    minimum_payment=payment.minimum_payment,
                    strategy_payment=payment.strategy_payment,
                    total_payment=payment.total_payment,
                    principal=payment.principal,
                    ending_balance=payment.ending_balance,
                )
                for payment in row.payments
            ],
            total_payment=row.total_payment,
            remaining_balance=row.remaining_balance,
        )
        for row in scenario.schedule
    ]
    return ScenarioResultResponse(
        strategy=scenario.strategy,
        payoff_order=list(scenario.payoff_order),
        extra_monthly_payment=scenario.extra_monthly_payment,
        monthly_payment_budget=scenario.monthly_payment_budget,
        schedule=schedule,
        total_interest=scenario.total_interest,
        total_paid=scenario.total_paid,
        months=scenario.months,
        payoff_date=schedule[-1].payment_date if schedule else None,
        months_saved=baseline_months - scenario.months,
        interest_saved=baseline_interest - scenario.total_interest,
    )


@projection_router.post("/scenarios", response_model=ScenarioComparisonResponse)
def project_scenarios(data: ScenarioProjectionRequest) -> ScenarioComparisonResponse:
    scenario_debts = tuple(
        debt.ScenarioDebt(
            debt_id=item.debt_id,
            balance=item.balance,
            annual_rate_basis_points=item.annual_rate_basis_points,
            minimum_payment=item.minimum_payment,
        )
        for item in data.debts
    )
    custom_order = tuple(data.custom_order) if data.custom_order is not None else None
    try:
        projections = [
            debt.calculate_payoff_scenario(
                debts=scenario_debts,
                strategy=strategy,
                extra_monthly_payment=data.extra_monthly_payment,
                custom_order=custom_order if strategy == debt.PayoffStrategy.CUSTOM else None,
                max_months=data.max_months,
            )
            for strategy in data.strategies
        ]
    except debt.AmortizationError as error:
        raise _projection_error(error) from error

    baseline = max(
        projections,
        key=lambda item: (item.months, item.total_interest, item.strategy.value),
    )
    return ScenarioComparisonResponse(
        assumptions=_assumptions(data.max_months),
        comparison_baseline=baseline.strategy,
        scenarios=[
            _scenario_response(
                scenario,
                data.first_payment_date,
                baseline_months=baseline.months,
                baseline_interest=baseline.total_interest,
            )
            for scenario in projections
        ],
    )

File: frontend/src/api/debt.ts
export type LoanTerms = {
  id: string
  account_id: string
  principal: string
  annual_rate_basis_points: number
  minimum_payment: string
  term_months: number | null
  created_at: string
  updated_at: string
}

export type LoanTermsInput = Omit<LoanTerms, 'id' | 'created_at' | 'updated_at'>

export type LoanBalance = {
  id: string
  loan_terms_id: string
  as_of_date: string
  balance: string
  source: string
  created_at: string
  updated_at: string
}

export type LoanBalanceInput = Pick<LoanBalance, 'as_of_date' | 'balance' | 'source'>

export type ProjectionAssumptions = {
  apr_treatment: string
  periodic_rate: string
  compounding: string
  payment_timing: string
  currency_rounding: string
  final_payment: string
  maximum_months: number
  disclaimer: string
}

export type AmortizationPayment = {
  month: number
  payment_date: string
  starting_balance: string
  interest: string
  payment: string
  principal: string
  ending_balance: string
}

export type AmortizationProjection = {
  assumptions: ProjectionAssumptions
  payments: AmortizationPayment[]
  total_interest: string
  total_paid: string
  months: number
  payoff_date: string | null
  annual_rate_basis_points: number
  monthly_rate: string
}

export type AmortizationInput = {
  principal: string
  annual_rate_basis_points: number
  monthly_payment: string
  first_payment_date: string
}

type ApiErrorBody = { detail?: string }

async function request<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, init)
  if (!response.ok) {
    let message = 'The debt request failed.'
    try {
      const body = (await response.json()) as ApiErrorBody
      if (typeof body.detail === 'string') message = body.detail
    } catch {
      // Keep the generic message for non-JSON failures.
    }
    throw new Error(message)
  }
  if (response.status === 204) return undefined as T
  return (await response.json()) as T
}

const json = (method: string, body: unknown): RequestInit => ({
  method,
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify(body),
})

export const listLoans = () => request<LoanTerms[]>('/api/v1/loans')
export const createLoan = (input: LoanTermsInput) =>
  request<LoanTerms>('/api/v1/loans', json('POST', input))
export const updateLoan = (id: string, input: Partial<LoanTermsInput>) =>
  request<LoanTerms>(`/api/v1/loans/${id}`, json('PATCH', input))
export const deleteLoan = (id: string) =>
  request<void>(`/api/v1/loans/${id}`, { method: 'DELETE' })

export const listLoanBalances = (loanId: string) =>
  request<LoanBalance[]>(`/api/v1/loans/${loanId}/balances`)
export const createLoanBalance = (loanId: string, input: LoanBalanceInput) =>
  request<LoanBalance>(`/api/v1/loans/${loanId}/balances`, json('POST', input))
export const updateLoanBalance = (loanId: string, balanceId: string, input: LoanBalanceInput) =>
  request<LoanBalance>(
    `/api/v1/loans/${loanId}/balances/${balanceId}`,
    json('PATCH', input),
  )
export const deleteLoanBalance = (loanId: string, balanceId: string) =>
  request<void>(`/api/v1/loans/${loanId}/balances/${balanceId}`, { method: 'DELETE' })

export const projectAmortization = (input: AmortizationInput) =>
  request<AmortizationProjection>('/api/v1/debt/amortization', json('POST', input))
